pkg/common/cns-lib/volume/manager.go

```diff
@@ -608,3 +618,52 @@ func (m *defaultManager) QueryAllVolume(ctx context.Context, queryFilter cnstype
 	}
 	return res, err
 }
+
+// QueryVolumeInfo calls the CNS QueryVolumeInfo API and return a task, from which CnsQueryVolumeInfoResult is extracted
+func (m *defaultManager) QueryVolumeInfo(ctx context.Context, volumeIDList []cnstypes.CnsVolumeId) (*cnstypes.CnsQueryVolumeInfoResult, error) {
+	log := logger.GetLogger(ctx)
+	err := validateManager(ctx, m)
+	if err != nil {
+		return nil, err
+	}
+	// Set up the VC connection
+	err = m.virtualCenter.ConnectCns(ctx)
+	if err != nil {
+		log.Errorf("ConnectCns failed with err: %+v", err)
+		return nil, err
+	}
+	//Call the CNS QueryVolumeInfo
+	queryVolumeInfoTask, err := m.virtualCenter.CnsClient.QueryVolumeInfo(ctx, volumeIDList)
+	if err != nil {
+		log.Errorf("CNS QueryAllVolume failed from vCenter %q with err: %v", m.virtualCenter.Config.Host, err)
+		return nil, err
+	}
+
+	// Get the taskInfo
+	taskInfo, err := cns.GetTaskInfo(ctx, queryVolumeInfoTask)
+	if err != nil || taskInfo == nil {
+		log.Errorf("failed to get taskInfo for QueryVolumeInfo task from vCenter %q with err: %v", m.virtualCenter.Config.Host, err)
+		return nil, err
+	}
+	log.Infof("QueryVolumeInfo: volumeIDList: %v, opId: %q", volumeIDList, taskInfo.ActivationId)
+	// Get the task results for the given task
+	taskResult, err := cns.GetTaskResult(ctx, taskInfo)
+	if err != nil {
+		log.Errorf("unable to find the task result for QueryVolumeInfo task from vCenter %q with taskID %s and taskResult %v",
+			m.virtualCenter.Config.Host, taskInfo.Task.Value, taskResult)
+		return nil, err
+	}
+	if taskResult == nil {
+		log.Errorf("taskResult is empty for DeleteVolume task: %q, opID: %q", taskInfo.Task.Value, taskInfo.ActivationId)
+		return nil, errors.New("taskResult is empty")
+	}
+	volumeOperationRes := taskResult.GetCnsVolumeOperationResult()
+	if volumeOperationRes.Fault != nil {
+		msg := fmt.Sprintf("failed to Query volumes: %v, fault: %q, opID: %q", volumeIDList, spew.Sdump(volumeOperationRes.Fault), taskInfo.ActivationId)
+		log.Error(msg)
+		return nil, errors.New(msg)
+	}
+	volumeInfoResult := interface{}(taskResult).(*cnstypes.CnsQueryVolumeInfoResult)
```

Comment:
Question: CNS returns an array of QueryVolumeInfoResult (base class VolumeOperationResult) in VolumeOperationBatchResult. So, how come we are not dealing with arrays here?

Reply:
Base of `CnsQueryVolumeInfoResult` is `CnsVolumeOperationResult`.


```
type CnsVolumeOperationResult struct {
	types.DynamicData

	VolumeId CnsVolumeId                 `xml:"volumeId,omitempty"`
	Fault    *types.LocalizedMethodFault `xml:"fault,omitempty"`
}

type CnsQueryVolumeInfoResult struct {
	CnsVolumeOperationResult

	VolumeInfo BaseCnsVolumeInfo `xml:"volumeInfo,typeattr,omitempty"`
}
```


with one volume I got a response from the server as below. Let me supply multiple volumes and see if any binding correction is required.

```
client_test.go:222: Calling QueryVolumeInfo using: []types.CnsVolumeId{
            {
                Id: "182f0fe0-71da-4ff4-a1da-47dd202254f3",
            },
        }
    client_test.go:246: Sucessfully Queried Volumes. queryVolumeInfoTaskResult: &types.CnsQueryVolumeInfoResult{
            CnsVolumeOperationResult: types.CnsVolumeOperationResult{},
            VolumeInfo:               &types.CnsBlockVolumeInfo{
                CnsVolumeInfo:  types.CnsVolumeInfo{},
                VStorageObject: types.VStorageObject{
                    Config: types.VStorageObjectConfigInfo{
                        BaseConfigInfo: types.BaseConfigInfo{
                            Id: types.ID{
                                Id: "182f0fe0-71da-4ff4-a1da-47dd202254f3",
                            },
                            Name:                        "pvc-901e87eb-c2bd-11e9-806f-005056a0c9a0",
                            CreateTime:                  time.Now(),
                            KeepAfterDeleteVm:           types.NewBool(true),
                            RelocationDisabled:          types.NewBool(false),
                            NativeSnapshotSupported:     types.NewBool(false),
                            ChangedBlockTrackingEnabled: types.NewBool(false),
                            Backing:                     &types.BaseConfigInfoDiskFileBackingInfo{
                                BaseConfigInfoFileBackingInfo: types.BaseConfigInfoFileBackingInfo{
                                    BaseConfigInfoBackingInfo: types.BaseConfigInfoBackingInfo{
                                        Datastore: types.ManagedObjectReference{Type:"Datastore", Value:"datastore-66"},
                                    },
                                    FilePath:        "[vsanDatastore] 0c84bd5e-9acb-f2c9-785b-020a0779fc97/993e899e9e984c81bb9af778775c7a2e.vmdk",
                                    BackingObjectId: "1384bd5e-6685-97f3-ad74-020a0779fc97",
                                    Parent:          nil,
                                    DeltaSizeInMB:   0,
                                    KeyId:           (*types.CryptoKeyId)(nil),
                                },
                                ProvisioningType: "thin",
                            },
                            Iofilter: nil,
                        },
                        CapacityInMB:    5120,
                        ConsumptionType: []string{"disk"},
                        ConsumerId:      nil,
                    },
                },
            },
        }
```

Reply:
@SandeepPissay we have a bug in the CNS go binding for QueryVolumeInfo - https://github.com/vmware/govmomi/issues/2010

Current binding works with 1 volume, but I will take a look and fix it. 
I will revisit this logic in the driver, once go binding is fixed.